File: companion-ios/SentryKey/Import/GoogleAuthImport.swift
import Foundation

/// Imports accounts from Google Authenticator's `otpauth-migration://` export QR.
/// Hand-parses the base64 protobuf (no dependency) and Base32-encodes the raw
/// secrets into our format. Mirrors the Android GoogleAuthImport.
enum GoogleAuthImport {

    static func isMigrationURI(_ raw: String) -> Bool {
        raw.lowercased().hasPrefix("otpauth-migration://")
    }

    static func parse(_ raw: String) -> [TwoFactorAccount] {
        guard isMigrationURI(raw), let range = raw.range(of: "data=") else { return [] }
        var encoded = String(raw[range.upperBound...])
        if let amp = encoded.firstIndex(of: "&") { encoded = String(encoded[..<amp]) }
        guard let decoded = encoded.removingPercentEncoding,
              let data = Data(base64Encoded: decoded) else { return [] }
        return parsePayload([UInt8](data))
    }

    private static func parsePayload(_ bytes: [UInt8]) -> [TwoFactorAccount] {
        var reader = Reader(data: bytes)
        var out: [TwoFactorAccount] = []
        while reader.hasMore {
            let (field, wire) = reader.tag()
            if field == 1 && wire == 2 {
                if let acc = parseOtpParameters(reader.bytes()) { out.append(acc) }
            } else {
                reader.skip(wire)
            }
        }
        return out
    }

    private static func parseOtpParameters(_ bytes: [UInt8]) -> TwoFactorAccount? {
        var reader = Reader(data: bytes)
        var secret: [UInt8]?
        var name = ""
        var issuer = ""
        while reader.hasMore {
            let (field, wire) = reader.tag()
            switch (field, wire) {
            case (1, 2): secret = reader.bytes()
            case (2, 2): name = String(decoding: reader.bytes(), as: UTF8.self)
            case (3, 2): issuer = String(decoding: reader.bytes(), as: UTF8.self)
            default: reader.skip(wire)
            }
        }
        guard let raw = secret else { return nil }
        let base32 = base32Encode(raw)
        if base32.isEmpty { return nil }

        let label: String
        if !issuer.isEmpty, !name.isEmpty, !name.localizedCaseInsensitiveContains(issuer) {
            label = "\(issuer):\(name)"
        } else if !name.isEmpty {
            label = name
        } else if !issuer.isEmpty {
            label = issuer
        } else {
            label = "Account"
        }
        return TwoFactorAccount(label: label, secret: base32)
    }

    static func base32Encode(_ data: [UInt8]) -> String {
        let alphabet = Array("ABCDEFGHIJKLMNOPQRSTUVWXYZ234567")
        var out = ""
        var buffer = 0
        var bits = 0
        for byte in data {
            buffer = (buffer << 8) | Int(byte)
            bits += 8
            while bits >= 5 {
                bits -= 5
                out.append(alphabet[(buffer >> bits) & 0x1F])
            }
            buffer &= (1 << bits) - 1
        }
        if bits > 0 {
            out.append(alphabet[(buffer << (5 - bits)) & 0x1F])
        }
        return out
    }

    /// Minimal protobuf wire-format reader (varint + length-delimited).
    private struct Reader {
        let data: [UInt8]
        var pos = 0
        var hasMore: Bool { pos < data.count }

        mutating func varint() -> Int {
            var result = 0
            var shift = 0
            while pos < data.count {
                let b = Int(data[pos]); pos += 1
                result |= (b & 0x7F) << shift
                if b & 0x80 == 0 { break }
                shift += 7
            }
            return result
        }

        mutating func tag() -> (Int, Int) {
            let t = varint()
            return (t >> 3, t & 0x7)
        }

        mutating func bytes() -> [UInt8] {
            let len = varint()
            let end = min(pos + len, data.count)
            let slice = Array(data[pos..<end])
            pos = end
            return slice
        }

        mutating func skip(_ wire: Int) {
            switch wire {
            case 0: _ = varint()
            case 1: pos += 8
            case 2: pos += varint()
            case 5: pos += 4
            default: break
            }
        }
    }
}

File: companion/app/src/main/java/com/example/sentrykey/GoogleAuthImport.kt
package com.example.sentrykey

import android.net.Uri
import android.util.Base64

/**
 * Imports accounts from Google Authenticator's "Export accounts" QR codes,
 * which encode `otpauth-migration://offline?data=<base64 protobuf>`.
 *
 * The protobuf is parsed by hand (no dependency). Schema (MigrationPayload):
 *   repeated OtpParameters otp_parameters = 1;
 *   OtpParameters { bytes secret=1; string name=2; string issuer=3; ... }
 * GA stores raw secret bytes, so we Base32-encode them into our format.
 */
object GoogleAuthImport {

    fun isMigrationUri(raw: String): Boolean =
        raw.startsWith("otpauth-migration://", ignoreCase = true)

    fun parse(raw: String): List<TwoFactorAccount> {
        if (!isMigrationUri(raw)) return emptyList()
        val marker = "data="
        val idx = raw.indexOf(marker)
        if (idx < 0) return emptyList()
        val dataParam = raw.substring(idx + marker.length).substringBefore('&')
        return try {
            val decoded = Uri.decode(dataParam)
            val bytes = Base64.decode(decoded, Base64.DEFAULT)
            parsePayload(bytes)
        } catch (e: Exception) {
            emptyList()
        }
    }

    private fun parsePayload(data: ByteArray): List<TwoFactorAccount> {
        val out = ArrayList<TwoFactorAccount>()
        val reader = ProtoReader(data)
        while (reader.hasMore()) {
            val (field, wire) = reader.readTag()
            if (field == 1 && wire == 2) {
                parseOtpParameters(reader.readBytes())?.let { out.add(it) }
            } else {
                reader.skip(wire)
            }
        }
        return out
    }

    private fun parseOtpParameters(data: ByteArray): TwoFactorAccount? {
        val reader = ProtoReader(data)
        var secret: ByteArray? = null
        var name = ""
        var issuer = ""
        while (reader.hasMore()) {
            val (field, wire) = reader.readTag()
            when {
                field == 1 && wire == 2 -> secret = reader.readBytes()
                field == 2 && wire == 2 -> name = String(reader.readBytes(), Charsets.UTF_8)
                field == 3 && wire == 2 -> issuer = String(reader.readBytes(), Charsets.UTF_8)
                else -> reader.skip(wire)
            }
        }
        val raw = secret ?: return null
        val base32 = base32Encode(raw)
        if (base32.isEmpty()) return null

        val label = when {
            issuer.isNotEmpty() && name.isNotEmpty() && !name.contains(issuer, ignoreCase = true) ->
                "$issuer:$name"
            name.isNotEmpty() -> name
            issuer.isNotEmpty() -> issuer
            else -> "Account"
        }
        return TwoFactorAccount(label, base32)
    }

    /** Minimal protobuf wire-format reader (varint + length-delimited). */
    private class ProtoReader(private val data: ByteArray) {
        private var pos = 0
        fun hasMore() = pos < data.size

        fun readVarint(): Long {
            var result = 0L
            var shift = 0
            while (pos < data.size) {
                val b = data[pos++].toInt() and 0xFF
                result = result or ((b and 0x7F).toLong() shl shift)
                if (b and 0x80 == 0) break
                shift += 7
            }
            return result
        }

        fun readTag(): Pair<Int, Int> {
            val tag = readVarint()
            return Pair((tag ushr 3).toInt(), (tag and 0x7L).toInt())
        }

        fun readBytes(): ByteArray {
            val len = readVarint().toInt()
            val end = (pos + len).coerceAtMost(data.size)
            val slice = data.copyOfRange(pos, end)
            pos = end
            return slice
        }

        fun skip(wire: Int) {
            when (wire) {
                0 -> readVarint()
                1 -> pos += 8
                2 -> pos += readVarint().toInt()
                5 -> pos += 4
            }
        }
    }
}

/** RFC 4648 Base32 encoder (no padding) — inverse of the app's decodeBase32. */
fun base32Encode(data: ByteArray): String {
    val alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"
    val sb = StringBuilder()
    var buffer = 0
    var bits = 0
    for (byte in data) {
        buffer = (buffer shl 8) or (byte.toInt() and 0xFF)
        bits += 8
        while (bits >= 5) {
            bits -= 5
            sb.append(alphabet[(buffer ushr bits) and 0x1F])
        }
        buffer = buffer and ((1 shl bits) - 1)
    }
    if (bits > 0) {
        sb.append(alphabet[(buffer shl (5 - bits)) and 0x1F])
    }
    return sb.toString()
}

/** Resolves a scanned QR to accounts: GA migration (many) or a single otpauth. */
fun parseScanResult(raw: String): List<TwoFactorAccount> =
    if (GoogleAuthImport.isMigrationUri(raw)) {
        GoogleAuthImport.parse(raw)
    } else {
        parseOtpauthUri(raw)?.let { listOf(it) } ?: emptyList()
    }
